src/db/sqlite/models.rs

```diff
@@ -0,0 +1,1170 @@
+    /// locks do SELECT ... LOCK IN SHARE MODE and write locks do SELECT
+    /// ... FOR UPDATE.
+    ///
+    /// In theory it would be possible to use serializable transactions rather
+    /// than explicit locking, but our ops team have expressed concerns about
+    /// the efficiency of that approach at scale.
+    pub fn lock_for_read_sync(&self, params: params::LockCollection) -> Result<()> {
+        let user_id = params.user_id.legacy_id as i64;
+        let collection_id =
+            self.get_collection_id(&params.collection)
+                .or_else(|e| match e.kind() {
+                    // If the collection doesn't exist, we still want to start a
+                    // transaction so it will continue to not exist.
+                    DbErrorKind::CollectionNotFound => Ok(0),
+                    _ => Err(e),
+                })?;
+        // If we already have a read or write lock then it's safe to
+        // use it as-is.
+        if self
+            .session
+            .borrow()
+            .coll_locks
+            .get(&(user_id as u32, collection_id))
+            .is_some()
+        {
+            return Ok(());
+        }
+
+        // Lock the db
+        self.begin(false)?;
+        let modified = user_collections::table
+            .select(user_collections::modified)
+            .filter(user_collections::user_id.eq(user_id))
+            .filter(user_collections::collection_id.eq(collection_id))
+            //            .lock_in_share_mode()
+            .first(&self.conn)
+            .optional()?;
+        if let Some(modified) = modified {
+            let modified = SyncTimestamp::from_i64(modified)?;
+            self.session
+                .borrow_mut()
+                .coll_modified_cache
+                .insert((user_id as u32, collection_id), modified); // why does it still expect a u32 int?
+        }
+        // XXX: who's responsible for unlocking (removing the entry)
+        self.session
+            .borrow_mut()
+            .coll_locks
+            .insert((user_id as u32, collection_id), CollectionLock::Read);
+        Ok(())
+    }
+
+    pub fn lock_for_write_sync(&self, params: params::LockCollection) -> Result<()> {
+        let user_id = params.user_id.legacy_id as i64;
+        let collection_id = self.get_or_create_collection_id(&params.collection)?;
+        if let Some(CollectionLock::Read) = self
+            .session
+            .borrow()
+            .coll_locks
+            .get(&(user_id as u32, collection_id))
+        {
+            Err(DbError::internal("Can't escalate read-lock to write-lock"))?
+        }
+
+        // Lock the db
+        self.begin(true)?;
+        let modified = user_collections::table
+            .select(user_collections::modified)
+            .filter(user_collections::user_id.eq(user_id))
+            .filter(user_collections::collection_id.eq(collection_id))
+            //            .for_update()
```

Comment:
Same as above, I don't know how to make this work in sqlite or if it is needed.

As far as I understood it (from a very brief read) this works in tandem with the line referenced above.